firmware/src/services/iot/cloud/mqtt_packetPopulation/mqtt_iothub_packetPopulate.c

```diff
@@ -0,0 +1,284 @@
+ * E.g.: For a particular topic
+ *       mchp/mySubscribedTopic/myDetailedPath
+ *       Sample publish handler function  = void handlePublishMessage(uint8_t *topic, uint8_t *payload)
+ */
+publishReceptionHandler_t imqtt_publishReceiveCallBackTable[MAX_NUM_TOPICS_SUBSCRIBE];
+
+void MQTT_CLIENT_iothub_publish(uint8_t* topic, uint8_t* payload, uint16_t payload_len, int qos)
+{
+    uint16_t packet_id = 0;
+    int      qos_value = 0;
+
+    if (topic == NULL)
+    {
+        debug_printError("  HUB: %s() missing PUBLISH topic");
+        return;
+    }
+
+    if (qos == 1)
+    {
+        qos_value = 1;
+        packet_id = ++packet_identifier;
+    }
+
+    //debug_printGood("  HUB: PUBLISH with ID %d", packet_identifier);
+
+    mqttPublishPacket cloudPublishPacket;
+    // Fixed header
+    cloudPublishPacket.publishHeaderFlags.duplicate = 0;
+    cloudPublishPacket.publishHeaderFlags.qos       = qos_value;
+    cloudPublishPacket.publishHeaderFlags.retain    = 0;
+
+    if (qos_value == 1)
+    {
+        cloudPublishPacket.packetIdentifierLSB = packet_id & 0xff;
+        cloudPublishPacket.packetIdentifierMSB = packet_id >> 8;
+    }
+
+    cloudPublishPacket.topic = topic;
+
+    // Payload
+    cloudPublishPacket.payload       = payload;
+    cloudPublishPacket.payloadLength = payload_len;
+
+    if (MQTT_CreatePublishPacket(&cloudPublishPacket) != true)
+    {
+        debug_printError("  HUB: MQTT_CLIENT_iothub_publish() failed");
+    }
+
+    return;
+}
+
+void MQTT_CLIENT_iothub_receive(uint8_t* data, uint16_t len)
+{
+    mqttHeaderFlags* mqtt_data = (mqttHeaderFlags*)data;
+
+    if (mqtt_data->controlPacketType == PUBACK)
+    {
+#ifdef DEBUG_PUBACK
+        mqttPubackPacket* mqtt_puback = (mqttPubackPacket*)data;
+        uint16_t          identifier  = mqtt_puback->packetIdentifierLSB << 8 | mqtt_puback->packetIdentifierMSB;
+
+        debug_printGood("  HUB: Received PUBACK for Packet ID %d", identifier);
+#endif
+    }
+    MQTT_GetReceivedData(data, len);
+}
+
+void MQTT_CLIENT_iothub_puback_callback(mqttPubackPacket* data)
+{
+#ifdef DEBUG_PUBACK
+    debug_printGood("  HUB: %s() Packet %d", __FUNCTION__, (uint16_t)(data->packetIdentifierMSB << 8 | data->packetIdentifierLSB));
+#endif
+    return;
+}
+
+void MQTT_CLIENT_iothub_connect(char* device_id)
+{
+    az_result     rc;
+    size_t        mqtt_connect_username_len;
+    const az_span hub_hostname_span    = az_span_create_from_str(hub_hostname);
+    const az_span device_id_span_local = az_span_create_from_str(device_id);
+    device_id_span                     = AZ_SPAN_FROM_BUFFER(device_id_buffer);
+
+    debug_printInfo("  HUB: Sending MQTT CONNECT to '%s'", hub_hostname);
+
+    LED_SetCloud(LED_INDICATOR_PENDING);
+
+#ifdef IOT_PLUG_AND_PLAY_MODEL_ID
+    az_span_copy(device_id_span, device_id_span_local);
+
+    device_id_span = az_span_slice(device_id_span, 0, az_span_size(device_id_span_local));
+#else
+    device_id_span = NULL;
+#endif
```

Comment:
I'm not sure this section is right. You would need a device_id regardless of plug and play or not. Also I think line 166 would fail compilation since it's a struct not a pointer. Bare minimum that should be
```suggestion
#ifdef IOT_PLUG_AND_PLAY_MODEL_ID
    az_span_copy(device_id_span, device_id_span_local);

    device_id_span = az_span_slice(device_id_span, 0, az_span_size(device_id_span_local));
#else
    device_id_span = AZ_SPAN_EMPTY;
#endif
```

Reply:
refactored code to add Non PnP Client code path (Use IoT Hub client vs. PnP Client).

[Commit](https://github.com/daisukeiot/Microchip-SAM-IoT-Wx/commit/cbb4ec08719089dd0108d3fad0be651692805f2c)